### PR TITLE
Fix codegen for schema-based NodeJS SDKs

### PR DESCRIPTION
**Pull request by lblackstone**
Related to pulumi/pulumi-kubernetes#1053

### CODE REVIEW

**Comment by lblackstone** [comment]
@pgavlin This is RFR

File: pkg/codegen/nodejs/gen.go [hunk 1]
```diff
@@ -864,6 +875,7 @@ func (mod *modContext) genConfig(w io.Writer, variables []*schema.Property) erro
 		printComment(w, p.Comment, "", "")
 
 		configFetch := fmt.Sprintf("__config.%s(\"%s\")", getfunc, p.Name)
+		// TODO: handle ConstValues
```
Comment (pgavlin) [inline pkg/codegen/nodejs/gen.go:878]: Could you file an issue for this?

File: pkg/codegen/nodejs/gen.go [hunk 2]
```diff
@@ -1125,12 +1139,21 @@ func (mod *modContext) genIndex(exports []string) string {
...
-		children = append(children, mod.mod)
+		child := mod.mod
+		// Trim version suffix from child modules. Nested versions will have their own index.ts file.
+		if versionSuffix.MatchString(child) {
+			if i := strings.LastIndex(child, "/"); i > 0 {
+				child = child[:i]
+			}
+		}
+		children.Add(child)
 	}
```
Comment (pgavlin) [inline pkg/codegen/nodejs/gen.go:1153]: This seems pretty k8s-specific. What's the concrete effect of this code?

File: pkg/codegen/nodejs/gen.go [hunk 3]
```diff
@@ -380,7 +384,11 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) error {
 		fmt.Fprintf(w, "     *\n")
 		fmt.Fprintf(w, "     * @param name The _unique_ name of the resulting resource.\n")
 		fmt.Fprintf(w, "     * @param id The _unique_ provider ID of the resource to lookup.\n")
-		fmt.Fprintf(w, "     * @param state Any extra arguments used during the lookup.\n")
+		// TODO: add k8s specific ID format doc?
```
[the remark this answers is not in the document]
Comment (lblackstone) [reply pkg/codegen/nodejs/gen.go:387]: https://github.com/pulumi/pulumi/issues/4754

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (lblackstone) [reply pkg/codegen/nodejs/gen.go:878]: https://github.com/pulumi/pulumi/issues/4755

File: pkg/codegen/nodejs/gen.go (hunk 2 above)
Comment (lblackstone) [reply pkg/codegen/nodejs/gen.go:1153]: It changes the import paths. 
```diff
-import * as apps from "./apps";
-export {apps};
+import * as apps/v1 from "./apps/v1";
+import * as apps/v1beta1 from "./apps/v1beta1";
+import * as apps/v1beta2 from "./apps/v1beta2";
+export {apps/v1, apps/v1beta1, apps/v1beta2};
```

Comment (pgavlin) [reply pkg/codegen/nodejs/gen.go:1150]: Got it. Can we put it under the k8s compatibility mode?

File: pkg/codegen/nodejs/gen.go [hunk 4]
```diff
@@ -1136,12 +1147,24 @@ func (mod *modContext) genIndex(exports []string) string {
...
 	for _, mod := range mod.children {
-		children = append(children, mod.mod)
+		child := mod.mod
+		if mod.compatibility == kubernetes20 {
+			// Trim version suffix from child modules. Nested versions will have their own index.ts file.
+			if match, _ := regexp.MatchString(`/v\d+((alpha|beta)\d+)?`, child); match {
+				if i := strings.LastIndex(child, "/"); i > 0 {
+					child = child[:i]
+				}
+			}
```
Comment (pgavlin) [inline pkg/codegen/nodejs/gen.go:1160]: I'd suggest compiling the regex and storing it as a package variable (also, if it's possible to anchor it to the end of the string, I'd add that):

```go
var k8sVersionRegex = regexp.MustCompile(`/v\d+((alpha|beta)\d+)?$`)
```

Then you can use `FindStringIndex` and avoid `strings.LastIndex`:

```suggestion
			if match := k8sVersionRegex.FindStringIndex(child); len(match) != 0 {
				child = child[:match[0]]
			}
```

File: pkg/codegen/nodejs/gen.go [hunk 5]
```diff
@@ -1153,10 +1156,8 @@ func (mod *modContext) genIndex(exports []string) string {
 		child := mod.mod
 		if mod.compatibility == kubernetes20 {
 			// Trim version suffix from child modules. Nested versions will have their own index.ts file.
-			if match, _ := regexp.MatchString(`/v\d+((alpha|beta)\d+)?`, child); match {
-				if i := strings.LastIndex(child, "/"); i > 0 {
-					child = child[:i]
-				}
+			if match := k8sVersionRegex.FindStringIndex(child); len(match) != 0 {
+				child = child[:match[0]-1] // -1 required to remove the leading '/'
```
Comment (pgavlin) [inline pkg/codegen/nodejs/gen.go:1160]: The `/` was in the regex before, which should obviate the need for the `-1`--was it necessary to remove the `/` for correctness?

Comment (lblackstone) [reply pkg/codegen/nodejs/gen.go:1160]: I did that so the name lined up better with what it was matching, but I could think up a new name if that makes more sense.

Comment (pgavlin) [reply pkg/codegen/nodejs/gen.go:1160]: It is safer to have the `/` in the regex, as the code that consumes the result is assuming its presence. I'd change the name of the regex.

**Pull request merged**